utf8 to utf16 conversion and utf16 to ascii

romainbrenguier opened the pull request:
Adds functions to `util/unicode.cpp` to convert from utf8 strings to utf16, both for little and big endian encodings and a function to pretty print utf16 characters in ascii.
These conversions are useful for the string refinement because utf16 is used by java.
Note that codecvt does not come automatically with old version of gcc, so I saw compilation problems with gcc-4.8 (Ubuntu 16.04), but none with gcc-6.

tautschnig commented on src/util/unicode.cpp:
```diff
@@ -256,8 +256,14 @@ const char **narrow_argv(int argc, const wchar_t **argv_wide)
 }
 
-std::wstring utf8_to_utf16(const std::string& in)
+std::wstring utf8_to_utf16be(const std::string& in)
```
It may be acceptable to spell out big_endian (and, below, little_endian). Don't make others decipher names.

tautschnig commented on src/util/unicode.cpp:
```diff
@@ -256,8 +256,14 @@ const char **narrow_argv(int argc, const wchar_t **argv_wide)
+
+std::wstring utf8_to_utf16le(const std::string& in)
+{
+  std::wstring_convert<std::codecvt_utf8_utf16<wchar_t,0x10ffffUL,std::codecvt_mode::little_endian> > converter;
```
Is there some macro or named constant that can be used in place of the hex value? Else it would be nice to have some comment explaining why it's exactly this bit pattern.

Also, the line should be broken somewhere before the 80th (or even 70th) character.

romainbrenguier replied:
Unfortunately I couldn't find such a constant, this is in fact the default for this template, hopefully this commit https://github.com/diffblue/cbmc/pull/372/commits/af792fe4d7e7f658317c1167457f31bc9b88278e makes it a bit clearer

tautschnig commented on src/util/unicode.cpp:
```diff
@@ -267,3 +267,22 @@ std::wstring utf8_to_utf16le(const std::string& in)
   std::wstring_convert<std::codecvt_utf8_utf16<wchar_t,0x10ffffUL,std::codecvt_mode::little_endian> > converter;
   return converter.from_bytes(in);
 }
```
Unrelated to the code: the commit message should be cleaned, we don't really care about that conflict anymore.

tautschnig commented on src/util/unicode.cpp:
```diff
@@ -267,3 +267,22 @@ std::wstring utf8_to_utf16le(const std::string& in)
   return converter.from_bytes(in);
 }
+
+std::string utf16le_to_ascii(const std::wstring& in)
```
Is this function missing a declaration in unicode.h?

tautschnig commented on src/util/unicode.cpp:
```diff
@@ -267,3 +267,22 @@ std::wstring utf8_to_utf16le(const std::string& in)
+  std::locale loc;
+  for(const auto c : in)
+  {
+    if(c <= 255 && isprint(c,loc))
```
No whitespace around `<=' but do add whitespace after ","

tautschnig commented on src/util/unicode.cpp:
```diff
@@ -267,3 +267,22 @@ std::wstring utf8_to_utf16le(const std::string& in)
+    {
+      result+="\\u";
+      char hex[5];
+      sprintf(hex,"%04x",(wchar_t)c);
```
Missing whitespace after ","

martin-cs commented:
May I suggest the test given at the end of this article:

https://mathiasbynens.be/notes/javascript-unicode

HTH

romainbrenguier commented:
@martin-cs The following test is handled correctly by the solver:

```
class test
{
    public static void main(String[] args)
    {
	String s="💩Iñtërnâtiônàlizætiøn☃";
	System.out.println(s.length());
	System.out.println(s.codePointAt(0));
	assert(s.length() == 23);
	assert(s.codePointAt(0) == 128169);
    }
}
```

Note however that the current solver is not able to check the same program for codePoint other than 0 (correspondence between code point and the actual index can be complicated in general).

martin-cs commented:
Cool; just thought that it might be a good test to add to find awkward
issues with multi-byte UTF16.

tautschnig commented:
@romainbrenguier Is the test you mentioned already part of some commit series? If it isn't it would be nice if it were added as a "FUTURE" test to this pull request (and then changed to "CORE" once all parts of the string solver are merged).

romainbrenguier commented:
The following test is already marked as future in CBMC: https://github.com/diffblue/cbmc/blob/master/regression/strings/java_code_point/test_code_point.java and it illustrates the same feature. But I can still add the "💩Iñtërnâtiônàlizætiøn☃" test, should it be part of this pull request or another one?

romainbrenguier commented on src/util/unicode.cpp:
```diff
@@ -253,3 +255,37 @@ const char **narrow_argv(int argc, const wchar_t **argv_wide)
+std::wstring utf8_to_utf16_little_endian(const std::string& in)
+{
+  const std::codecvt_mode mode=std::codecvt_mode::little_endian;
+  const unsigned long maxcode=0x10ffff;
```
I added a comment and a reference to the c++ documentation: https://github.com/diffblue/cbmc/pull/372/commits/ccbe6f77930851e2ca27986c0b91f1210c28ec3e

tautschnig commented:
Hmm, please hold off on merging this one: it seems further fixes for the travis integration of Clang are necessary in order to support codecvt.

romainbrenguier commented:
@tautschnig I squashed the commits and added the modification to .travis.yml you asked for https://github.com/diffblue/cbmc/pull/372/commits/c7a101ab1e0333a76f0b6ed20d4e414e6f5e451d , however I'm not sure the compilation with clang++ works

tautschnig commented on .travis.yml:
```diff
@@ -28,16 +28,16 @@ matrix:
             - clang-3.7
```
It seems possibly that we'll need an explicit
`- libc++-dev`
line in here.

tautschnig commented on .travis.yml:
```diff
@@ -28,16 +28,16 @@ matrix:
     - os: osx
       compiler: clang
-      env: COMPILER=clang++
+      env: COMPILER="clang++ -stdlib=libc++"
```
That change isn't necessary as libc++ is the default for Mac OS.

tautschnig commented:
If anyone has a Ubuntu system with Clang (3.7, ideally) at hand and could help debug the issues this would be much appreciated. At present, we are blocked as merging this work would break our travis set up.

romainbrenguier commented:
@tautschnig I tried the failing command on ubuntu with clang++-3.7 and reproduced the error gotten by travis but after installing libstd++-dev the error disappeared, so I don't know what's happening...

tautschnig commented:
@romainbrenguier Can you include in this pull request a travis config that matches the steps you took to make things work? It seems that installing libc++-dev is the wrong library (you stated libstdc++-dev works), and probably the compiler flags need to match as well. Thanks!!!

romainbrenguier commented:
@tautschnig sorry I meant libc++-dev , the travis file already mention the library I installed

romainbrenguier commented:
Using `-stdlib=libstdc++` instead of `-stdlib=libc++` makes the problem with `cctype` disappear put now there is a problem with `codecvt`

kroening commented:
Did you also try Visual Studio?

If C++ library support for the conversion isn't working yet, we might have to put our own implementation in.

smowton commented:
Note: Vojtech has asked me to look into the outstanding problems here. AFAIK they are:
1. figure out how and why this can be made to work with Clang (and specifically libc++, or libstd++ too?)
2. Similarly the VS versions we target

tautschnig commented:
@smowton In case you need confirmation: yes, I believe you have summarised the problems nicely.

smowton commented:
Confirming Romain's findings:

* G++ 5: success
* Clang++-3.7 / libstdc++-4.9: fails (codecvt not implemented)
* Clang++-3.7 / libstdc++-5: success
* Clang++-3.7 / libc++: success on my local machine, but not easy to do under Travis (there is no libc++ package for Ubuntu 12.04, which Travis runs in its VMs per default. We could switch to 14.04 per https://docs.travis-ci.com/user/ci-environment/ if we wanted to CI this case). Looking at the SVN history for libc++ / codecvt this seems likely to work back to around 2013.
* Visual Studio: success according to AppVeyor, and according to MSDN this should also work back to VS 2010 (according to the `COMPILING` doc we only support back to 2013 anyhow)

smowton commented:
https://github.com/diffblue/cbmc/pull/541 is a probable fix (waiting for Travis + AppVeyor)

smowton commented:
Fixed one other VS problem-- it doesn't have snprintf until VS 2015. Replaced with iostreams equivalent in #541.
